phpPackages.mkDerivation: use lib.{fix,extends} instead of //

Artturin opened the pull request:
// on a overrideAttrs'able works badly

here we use lib.fix and lib.extends instead of overrideAttrs to have
less of a performance impact

support added in
https://github.com/NixOS/nixpkgs/pull/119942

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

jtojnar commented on pkgs/development/php-packages/grumphp/default.nix:
```diff
@@ -1,10 +1,10 @@
 { mkDerivation, fetchurl, makeWrapper, lib, php }:
-mkDerivation rec {
+mkDerivation(finalAttrs: {
```
Function arguments should be separated by a space.

jtojnar replied:
Also, this should be `previousAttrs`.

jtojnar commented on pkgs/top-level/php-packages.nix:
```diff
@@ -53,10 +53,10 @@ lib.makeScope pkgs.newScope (self: with self; {
-  mkDerivation = { pname, ... }@args: pkgs.stdenv.mkDerivation (args // {
-    pname = "php-${pname}";
+  mkDerivation = attrs: (pkgs.stdenv.mkDerivation attrs).overrideAttrs(previousAttrs: {
+    name = "php-${previousAttrs.pname}-${previousAttrs.version}";
```
I do not really like this much – `name` is terrible concept and IMO should not exist other than as a label for store path for easier debugging. As such, I would like to hide it from Nixpkgs as much as possible. But yeah, it is not really clear who should be responsible for it, then.

But I guess if we set it here, it should probably refer to `finalAttrs` – if someone updates a version with `overrideAttrs`, the store path name should reflect that.

jtojnar reviewed (commented):
I love https://github.com/NixOS/nixpkgs/pull/119942 but I worry it might be too soon to make use of it. PRs like that have been reverted several times in the past.

jtojnar commented on pkgs/top-level/php-packages.nix:
```diff
@@ -53,10 +53,10 @@ lib.makeScope pkgs.newScope (self: with self; {
   # with how buildPecl does it and make the file easier to overview.
-  mkDerivation = { pname, ... }@args: pkgs.stdenv.mkDerivation (args // {
-    pname = "php-${pname}";
+  mkDerivation = attrs: (pkgs.stdenv.mkDerivation attrs).overrideAttrs(finalAttrs: _: {
```
```suggestion
  mkDerivation = attrs: (pkgs.stdenv.mkDerivation attrs).overrideAttrs (finalAttrs: _: {
```

jtojnar replied:
Still missing the space between function and argument.

FRidh commented:
Please hold off with making use of it in the weeks prior to branch off.

drupol commented:
What's the status of this PR now ?

Artturin commented:
It should use lib.fix and lib.extends instead of .overrideAttrs as those have less performance impact

Like i did here https://github.com/NixOS/nixpkgs/pull/211685/files#diff-b722c446a7a9291cb6e94d163d58351951ab9293999a54429883b5982fc8b1d6R19

drupol commented:
> It should use lib.fix and lib.extends instead of .overrideAttrs as those have less performance impact

Can you tell me a bit more about `it` ? What files are you talking about?

Also, do you plan to update your PR ?

Artturin committed:
phpPackages.mkDerivation: use lib.{fix,extends} instead of //

// on a overrideAttrs'able works badly

here we use lib.fix and lib.extends instead of overrideAttrs to have
less of a performance impact

drupol commented on pkgs/development/php-packages/grumphp/default.nix:
```diff
@@ -1,11 +1,11 @@
 { mkDerivation, fetchurl, makeWrapper, lib, php }:
 
-mkDerivation rec {
+mkDerivation (finalAttrs: {
```
So, if I understand correctly, we have to do that for all the packages in `pkgs/development/php-packages/` ?

Artturin replied:
No, its just to prove https://nixos.org/manual/nixpkgs/stable/#mkderivation-recursive-attributes works in packages using this mkDerivation.
